src/EditorFeatures/Test/Collections/IntervalTreeTests.cs

```diff
@@ -345,3 +349,67 @@ private static ISet<T> Set<T>(params T[] values)
         => new List<T>(values);
 }
+
+public sealed class BinaryIntervalTreeTests : IntervalTreeTests
+{
+    private protected override IEnumerable<IIntervalTree<Tuple<int, int, string>>> CreateTrees(IEnumerable<Tuple<int, int, string>> values)
+    {
+        yield return BinaryIntervalTree.Create(new TupleIntrospector<string>(), values);
+    }
+
+    private protected override bool HasIntervalThatIntersectsWith(IIntervalTree<Tuple<int, int, string>> tree, int position)
+    {
+        return ((BinaryIntervalTree<Tuple<int, int, string>>)tree).Algorithms.HasIntervalThatIntersectsWith(position, new TupleIntrospector<string>());
+    }
+
+    private protected override ImmutableArray<Tuple<int, int, string>> GetIntervalsThatIntersectWith(IIntervalTree<Tuple<int, int, string>> tree, int start, int length)
+    {
+        return ((BinaryIntervalTree<Tuple<int, int, string>>)tree).Algorithms.GetIntervalsThatIntersectWith(start, length, new TupleIntrospector<string>());
+    }
+
+    private protected override ImmutableArray<Tuple<int, int, string>> GetIntervalsThatOverlapWith(IIntervalTree<Tuple<int, int, string>> tree, int start, int length)
+    {
+        return ((BinaryIntervalTree<Tuple<int, int, string>>)tree).Algorithms.GetIntervalsThatOverlapWith(start, length, new TupleIntrospector<string>());
+    }
+}
+
+public sealed class FlatArrayIntervalTreeTests : IntervalTreeTests
+{
+    private protected override IEnumerable<IIntervalTree<Tuple<int, int, string>>> CreateTrees(IEnumerable<Tuple<int, int, string>> values)
+    {
+        yield return FlatArrayIntervalTree<Tuple<int, int, string>>.CreateFromUnsorted(new TupleIntrospector<string>(), new SegmentedList<Tuple<int, int, string>>(values));
+    }
+
+    private protected override bool HasIntervalThatIntersectsWith(IIntervalTree<Tuple<int, int, string>> tree, int position)
+    {
+        return ((FlatArrayIntervalTree<Tuple<int, int, string>>)tree).Algorithms.HasIntervalThatIntersectsWith(position, new TupleIntrospector<string>());
+    }
+
+    private protected override ImmutableArray<Tuple<int, int, string>> GetIntervalsThatIntersectWith(IIntervalTree<Tuple<int, int, string>> tree, int start, int length)
+    {
+        return ((FlatArrayIntervalTree<Tuple<int, int, string>>)tree).Algorithms.GetIntervalsThatIntersectWith(start, length, new TupleIntrospector<string>());
+    }
+
+    private protected override ImmutableArray<Tuple<int, int, string>> GetIntervalsThatOverlapWith(IIntervalTree<Tuple<int, int, string>> tree, int start, int length)
+    {
+        return ((FlatArrayIntervalTree<Tuple<int, int, string>>)tree).Algorithms.GetIntervalsThatOverlapWith(start, length, new TupleIntrospector<string>());
+    }
+
+    private readonly struct Int32IntervalIntrospector : IIntervalIntrospector<int>
+    {
+        public TextSpan GetSpan(int value)
+            => new(value, 0);
+    }
+
+    [Fact]
+    public void TestProperBalancing()
+    {
+        for (var i = 0; i < 3000; i++)
```

Comment:
Hammers a ton of configurations of interval trees, ensuring that our creation of the complete binary tree is correct.